service/vpn_key_rotation_service.go

```diff
@@ -0,0 +1,305 @@
+			cl, err := v.listClientPairGateway(workerSliceGatewaysList, gateway.Spec.RemoteGatewayConfig.GatewayName)
+			if err != nil {
+				return err
+			}
+			// construct clustermap
+			clusterCidr := util.FindCIDRByMaxClusters(s.Spec.MaxClusters)
+			completeResourceName := fmt.Sprintf(util.LabelValue, util.GetObjectKind(s), s.GetName())
+			ownershipLabel := util.GetOwnerLabel(completeResourceName)
+			workerSliceConfigs, err := v.wscs.ListWorkerSliceConfigs(ctx, ownershipLabel, s.Namespace)
+			if err != nil {
+				return err
+			}
+			clusterMap := v.wscs.ComputeClusterMap(s.Spec.Clusters, workerSliceConfigs)
+			// contruct gw address
+			gatewayAddresses := v.wsgs.BuildNetworkAddresses(s.Spec.SliceSubnet, gateway.Spec.LocalGatewayConfig.ClusterName, gateway.Spec.RemoteGatewayConfig.ClusterName, clusterMap, clusterCidr)
+			// call GenerateCerts()
+			if err := v.wsgs.GenerateCerts(ctx, s.Name, s.Namespace, &gateway, cl, gatewayAddresses); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (v *VpnKeyRotationService) listWorkerSliceGateways(ctx context.Context, labels map[string]string) (*workerv1alpha1.WorkerSliceGatewayList, error) {
+	workerSliceGatewaysList := workerv1alpha1.WorkerSliceGatewayList{}
+	// list workerslicegateways
+	listOpts := []client.ListOption{
+		client.MatchingLabels(
+			labels,
+		),
+	}
+	if err := util.ListResources(ctx, &workerSliceGatewaysList, listOpts...); err != nil {
+		return nil, err
+	}
+	return &workerSliceGatewaysList, nil
+}
+
+// getSliceConfig
+func (v *VpnKeyRotationService) getSliceConfig(ctx context.Context, name, namespace string) (*controllerv1alpha1.SliceConfig, error) {
+	s := controllerv1alpha1.SliceConfig{}
+	found, err := util.GetResourceIfExist(ctx, types.NamespacedName{
+		Name:      name,
+		Namespace: namespace,
+	}, &s)
+	if err != nil {
+		return nil, err
+	}
+	if !found {
+		return nil, fmt.Errorf("sliceconfig %s not found", name)
+	}
+	return &s, nil
+}
+
+func (v *VpnKeyRotationService) listClientPairGateway(wl *workerv1alpha1.WorkerSliceGatewayList, clientGatewayName string) (*workerv1alpha1.WorkerSliceGateway, error) {
+	for _, gateway := range wl.Items {
+		if gateway.Name == clientGatewayName {
+			return &gateway, nil
+		}
+	}
+	return nil, fmt.Errorf("cannot find gateway %s", clientGatewayName)
+}
+
+// verifyAllJobsAreCompleted checks if all the jobs are in complete state
+func (v *VpnKeyRotationService) verifyAllJobsAreCompleted(ctx context.Context, sliceName string) bool {
+	jobs := batchv1.JobList{}
+	o := map[string]string{
+		"SLICE_NAME": sliceName,
+	}
+	listOpts := []client.ListOption{
+		client.MatchingLabels(
+			o,
+		),
+	}
+	if err := util.ListResources(ctx, &jobs, listOpts...); err != nil {
+		return false
+	}
+	for _, job := range jobs.Items {
+		for _, condition := range job.Status.Conditions {
+			if !(condition.Type == batchv1.JobComplete && condition.Status == corev1.ConditionTrue) {
+				return false
+			}
+		}
```

Comment:
What if the JobConditionType  is 	Suspended or Failed?
I think we need to handle the case so that we don't be in the infinite loop

Reply:
Any update?

Reply:
https://github.com/kubeslice/kubeslice-controller/pull/169